pennylane/devices/qubit/adjoint_jacobian.py

```diff
@@ -108,3 +116,138 @@ def adjoint_jacobian(tape: QuantumTape):  # pylint: disable=too-many-statements
 
     # must be 2-dimensional
     return tuple(tuple(np.array(j_) for j_ in j) for j in jac)
+
+
+def adjoint_jvp(tape: QuantumTape, tangents: Tuple[Number]):
+    """Implements the adjoint method outlined in
+    `Jones and Gacon <https://arxiv.org/abs/2009.02823>`__ to differentiate an input tape.
+
+    After a forward pass, the circuit is reversed by iteratively applying adjoint
+    gates to scan backwards through the circuit.
+
+    .. note::
+        The adjoint differentiation method has the following restrictions:
+
+        * Only expectation values are supported as measurements.
+
+        * Cannot differentiate with respect to observables.
+
+        * Observable being measured must have a matrix.
+
+    Args:
+        tape (.QuantumTape): circuit that the function takes the gradient of
+
+    Returns:
+        array or tuple[array]: the derivative of the tape with respect to trainable parameters.
+        Dimensions are ``(len(observables), len(trainable_params))``.
+    """
+    # Map wires if custom wire labels used
+    if set(tape.wires) != set(range(tape.num_wires)):
```

Comment:
do we not care about wire order here?

Reply:
I wonder if we should actually move wire mapping to `preprocess` and assume we have integer wires by the time we get to `execute`, `execute_and_compute_derivatives`, `compute_jvp`, etc. Then we don't need to perform this translation in a bunch of places.

Reply:
@timmysilv There's no user-provided wire order here, so it doesn't matter as long as the wire mappings among the kets, bras, and operators are consistent (which they are).

Reply:
Added test for this below.